pkg/curatedpackages/packagecontrollerclient.go

```diff
@@ -49,19 +60,102 @@ type PackageControllerClient struct {
+	//
+	// Skipping the wait is desirable for full cluster lifecycle use cases,
+	// where resource creation and error reporting are asynchronous in nature.
+	skipWaitForPackageBundle bool
+	// tracker creates k8s clients for workload clusters managed via full
+	// cluster lifecycle API.
+	clientBuilder ClientBuilder
+
+	// mu provides some thread-safety.
+	mu sync.Mutex
+}
+
+// ClientBuilder returns a k8s client for the specified cluster.
+type ClientBuilder interface {
+	GetClient(context.Context, types.NamespacedName) (client.Client, error)
 }
 
 type ChartInstaller interface {
-	InstallChart(ctx context.Context, chart, ociURI, version, kubeconfigFilePath, namespace, valueFilePath string, values []string) error
+	InstallChart(ctx context.Context, chart, ociURI, version, kubeconfigFilePath, namespace, valueFilePath string, skipCRDs bool, values []string) error
+}
+
+// ChartInstallationDeleter handles deleting chart installations.
+type ChartInstallationDeleter interface {
+	Delete(ctx context.Context, kubeconfigFilePath, installName, namespace string) error
+}
+
+// NewPackageControllerClientFullLifecycle creates a PackageControllerClient
+// for the Full Cluster Lifecycle controller.
+//
+// It differs because the CLI use case has far more information available at
+// instantiation, while the FCL use case has less information at
+// instantiation, and the rest when cluster creation is triggered.
+func NewPackageControllerClientFullLifecycle(logger logr.Logger, chartInstaller ChartInstaller, deleter ChartInstallationDeleter, kubectl KubectlRunner, clientBuilder ClientBuilder) *PackageControllerClient {
+	return &PackageControllerClient{
+		chartInstaller:           chartInstaller,
+		deleter:                  deleter,
+		kubectl:                  kubectl,
+		skipWaitForPackageBundle: true,
+		eksaRegion:               eksaDefaultRegion,
+		clientBuilder:            clientBuilder,
+	}
+}
+
+// EnableFullLifecycle wraps Enable to handle run-time arguments.
+//
+// This method fills in the gaps between the original CLI use case, where all
+// information is known at PackageControllerClient initialization, and the
+// Full Cluster Lifecycle use case, where there's limited information at
+// initialization.
+func (pc *PackageControllerClient) EnableFullLifecycle(ctx context.Context, log logr.Logger, clusterName, kubeConfig string, chart *releasev1.Image, registryMirror *registrymirror.RegistryMirror, options ...PackageControllerClientOpt) (err error) {
+	log.V(6).Info("enabling curated packages full lifecycle")
+	defer func(err *error) {
+		if err != nil && *err != nil {
+			log.Error(*err, "Enabling curated packages full lifecycle", "clusterName", clusterName)
+		} else {
+			log.Info("Success enabling curated packages full lifecycle")
+		}
+	}(&err)
+	pc.mu.Lock()
+	// This anonymous function ensures that the pc.mu is unlocked before
+	// Enable is called, preventing deadlocks in the event that Enable tries
```

Comment:
Not sure if I quite understand why we need to use a lock here. Just curious on the use case.

Reply:
Guillermo pointed out that this isn't thread safe. While we don't have any concurrent code at the moment, he asked that I ensure thread-safety if possible, as there are designs towards concurrency for the future. Since it was relatively simple to add, I did so. This is far from the only thing in our codebase that isn't thread-safe though. ;)